_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2334 @@
+  https://www.runatlantis.io/docs/security.html#mitigations[all the security best practices] of the platform.
+
+https://www.hashicorp.com/products/terraform/[Terraform Enterprise and Terraform Cloud] (self-hosted and SaaS, specialized)::
+  Terraform Cloud (SaaS) and Terraform Enterprise (self-hosted) are platforms provided by HashiCorp (the creator of
+  Terraform) that are optimized for Terraform workflows. Both services provide a web UI for remotely running `terraform
+  plan` and `terraform apply`, and can be integrated with VCS platforms to implement CI/CD workflows that remotely run
+  `terraform` in reaction to git. Additionally, Terraform Enterprise in particular provides the ability to manage
+  deployment policies and permissions as code. Note that Terraform Cloud and Enterprise are both optimized for running
+  `terraform`. Both do not support directly calling external binaries such as `terragrunt` or `docker` in a workflow.
+  Terraform Cloud in particular does not support the use of external binaries even when called from within `terraform`
+  using a `local-exec` provisioner or external data source. In terms of the threat model, Terraform Cloud and Enterprise
+  both implement respectable security practices, including end to end encryption backed by HashiCorp Vault, and you have
+  the option to run in your own data center with Terraform Enterprise to lock it down. Note that Terraform Enterprise
+  relies on API based integrations with the VCS platforms instead of webhooks, so it is not necessary to have a publicly
+  facing service when running in self-hosted mode.
+
+Here is a table that highlights some of the features of each CI platform and their strengths and weaknesses that can be
+useful in making tradeoffs:
+
+[cols="1h,2,2,2,2,2"]
+|===
+| |Jenkins |CircleCI |GitLab |Atlantis |TFE and TFC
+
+|Hosting
+|Self-hosted
+|SaaS
+|SaaS or Self-hosted
+|Self-hosted
+|SaaS or Self-hosted
+
+|Generic?
+|Yes
+|Yes
+|Yes
+|Specialized to terraform
+|Specialized to terraform
+
+|VCS integration
+|Yes, with plugins
+|Yes
+|Yes
+|Yes
+|Yes
+
+|Provides static IP addresses for IP whitelisting
+|Yes
+|No
+|Yes (self-hosted)
+|Yes
+|Yes (TFE)
+
+|Update commit statuses
+|Yes
+|Yes
+|Yes
+|Yes
+|Yes
+
+|Annotate pull requests
+|Requires custom scripting
+|Requires custom scripting
+|Requires custom scripting
+|Yes
+|Yes
+
+|Supports multiple infrastructure tools
+|Yes
+|Yes
+|Yes
+|Limited (terraform only; additional binaries can be installed, but can not be called directly)
+|No (terraform only; TFE can support additional binaries, but TFC does not)
+
+|===
+
+
+[[production_grade_design]]
+== Production-grade design
+
+With all the core concepts out of the way, let's now discuss how to configure a production-grade CI/CD workflow for
+infrastructure code, using a platform that looks something like this:
+
+.Architecture of platform for running Terraform/Terragrunt CI/CD workflows.
```

Comment:
I found this diagram to be a bit confusing.

First, in the CI Servers box, the CI icons on the left side in the corner and then multiple raised boxes didn't make sense to me. Perhaps just have one raised box labeled "CI server" and leave off the icons?

Next, why are there multiple Lambda functions, but an arrow to just one of them? Also, the "Invoke Lambda Functions" text is too small to read. 

In the ECS box, I'd leave off the "ECS Fargate" icon, then rotate the "ECS Fargate" text 90deg and increase font size.

Reply:
I actually wonder if CloudCraft is the best tool here. A flat 2D diagram might be more effective.

Reply:
Ok I updated the image given your feedback: 93034c5

What do you think?

> Perhaps just have one raised box labeled "CI server" and leave off the icons?
IMO, the icons are important because they give a visual cue as to what this CI server should be.